pandas/tests/extension/decimal/test_decimal.py

```diff
@@ -63,9 +64,23 @@ def data_for_grouping():
 class BaseDecimal(object):
 
     def assert_series_equal(self, left, right, *args, **kwargs):
-
-        left_na = left.isna()
-        right_na = right.isna()
+        def convert(x):
+            # need to convert array([Decimal(NaN)], dtype='object') to np.NaN
+            # because Series[object].isnan doesn't recognize decimal(NaN) as
```

Comment:
where does this come up now? e.g. whats an example

Reply:
```python
In [21]: import pandas as pd

In [22]: import pandas.util.testing as tm

In [23]: from pandas.tests.extension.decimal import to_decimal

In [24]: ser = pd.Series(to_decimal(['1.0', 'NaN']))

In [25]: tm.assert_series_equal(ser.astype(object), ser.astype(object))
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-25-5356adf50e72> in <module>
----> 1 tm.assert_series_equal(ser.astype(object), ser.astype(object))

~/sandbox/pandas/pandas/util/testing.py in assert_series_equal(left, right, check_dtype, check_index_type, check_series_type, check_less_precise, check_names, check_exact, check_datetimelike_compat, check_categorical, obj)
   1293                                      check_less_precise=check_less_precise,
   1294                                      check_dtype=check_dtype,
-> 1295                                      obj='{obj}'.format(obj=obj))
   1296
   1297     # metadata comparison

~/sandbox/pandas/pandas/_libs/testing.pyx in pandas._libs.testing.assert_almost_equal()
     64
     65
---> 66 cpdef assert_almost_equal(a, b,
     67                           check_less_precise=False,
     68                           bint check_dtype=True,

~/sandbox/pandas/pandas/_libs/testing.pyx in pandas._libs.testing.assert_almost_equal()
    178             msg = '{0} values are different ({1} %)'.format(
    179                 obj, np.round(diff * 100.0 / na, 5))
--> 180             raise_assert_detail(obj, msg, lobj, robj)
    181
    182         return True

~/sandbox/pandas/pandas/util/testing.py in raise_assert_detail(obj, message, left, right, diff)
   1080         msg += "\n[diff]: {diff}".format(diff=diff)
   1081
-> 1082     raise AssertionError(msg)
   1083
   1084

AssertionError: Series are different

Series values are different (50.0 %)
[left]:  [1.0, NaN]
[right]: [1.0, NaN]
```

We do the astype(object) to build the expected.

Reply:
I think this is a bug, and would solve the aboe i think.

```
In [7]: ser.astype(object).isna()
Out[7]: 
0    False
1    False
dtype: bool
```

Reply:
@TomAugspurger actually, can you explain why the current code in master is not working fine? Why do you need to convert to object? Because before, there was already the calls to `isna` to check NaNs and non-NaNs separately.

Reply:
this is my point. I think there is a bug somewhere here, e.g. isna is maybe not dispatching to the EA?

Reply:
To construct the expected in `test_untsack` base test, we do `ser.astype(object).untsack(level=level)`. That means we convert the Series[Decimal] to Series[object], which has the incorrect `isna`.

Reply:
So it's all about creating the expected result. When we go to do the final assert that the values match, we do

```python
            result = result.astype(object)

            self.assert_frame_equal(result, expected)
```

but `self.assert_frame_equal` will say that `Series([Decimal('NaN')], dtype='object')` isn't equal to itself, since it doesn't consider that value NA.

Reply:
Ah, OK, I somehow thought you were doing the `astype(object)` inside the assert testing machinery above, not in the actual expected result. Yes, that makes sense now. 

For me it is fine to keep this hack in here for now. In the end that is somehow the purpose of using the class instances for `assert_.._equal` so a specific EA can override it.

Reply:
I think the ``.astype(object)`` for Decimal is incorrect.

```
In [5]: ser.values
Out[5]: DecimalArray(array([Decimal('1.0'), Decimal('NaN')], dtype=object))

In [6]: ser.astype(object)
Out[6]: 
0    1.0
1    NaN
dtype: object

In [7]: ser.astype(object).values
Out[7]: array([Decimal('1.0'), Decimal('NaN')], dtype=object)
```

on the these should be converted to ``np.nan`` and not ``Decimal('NaN')`` I think as this is just a numpy array.

Reply:
object-dtype can store anything, including decimal objects. It'd be strange to only convert `Decimal("NaN")` to `np.nan`, and not `Decimal('1.0')` to 1.0, no?

Reply:
Yeah, this seems correct to me. 

Going to merge, if we need to further discuss this, we can do that in another issue (it's not really related anymore with actually fixing unstack)